File: apps/windows/TeslaSync.App/feature-views/SummaryStatsRow.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using TeslaSync.App.Components.DataDisplay;
using TeslaSync.App.Components.Feedback;
using TeslaSync.App.Components.Motion;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The native WinUI 3 <c>SummaryStatsRow</c> feature surface — a parity port of
/// web/src/features/admin/components/security-access/SummaryStatsRow.tsx. It is a pure presentational row:
/// assign a <see cref="Model"/> (the web <c>isSecure</c> / <c>lastLockChange</c> / <c>sentryUptime</c> /
/// <c>totalEvents</c> props plus the parent's <c>isLoading</c> flag) and it renders the web's responsive grid
/// (<c>grid-cols-1 sm:grid-cols-2 lg:grid-cols-4</c>) of four metric tiles. While the parent fetches it shows
/// four tokenized <see cref="TsSkeleton"/> tiles (web <c>&lt;Skeleton height={88}/&gt;</c>); once resolved it
/// shows the four <see cref="TsMetricCard"/> tiles (web <c>&lt;MetricCard&gt;</c>) — Current Status (secure ⇒
/// success / unsecure ⇒ danger), Last Lock Change (cyan, relative time), Sentry Uptime (blue, percent) and Total
/// Events (purple, count) — entering through a single <see cref="TsFadeIn"/> (the native mapping of the web
/// <c>&lt;FadeIn&gt;</c>, honouring reduce-motion). The view never performs HTTP; all branch selection, label and
/// value resolution and relative-time formatting happen in the WinUI-free <see cref="SummaryStatsRowProjection"/>.
/// Every string resolves through the i18n facade, each tile and the surface carry a Narrator name, and the
/// loading grid announces itself through a live region.
/// </summary>
public sealed partial class SummaryStatsRow : ContentControl
{
    private const double SkeletonTileHeight = 88; // web <Skeleton height={88}/>
    private const double TileGap = 16;            // web gap-4
    private const double SmBreakpoint = 600;      // web sm: — two columns at/above this width
    private const double LgBreakpoint = 960;      // web lg: — four columns at/above this width

    private readonly ILocalizer _localizer;
    private readonly SummaryStatsRowDiagnostics _diagnostics;

    private SummaryStatsRowModel _model;
    private bool _opened;
    private int _columns = SummaryStatsRowProjection.TileCount;

    /// <summary>Creates the surface over its i18n facade, an initial model, and (optional) diagnostics.</summary>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <param name="model">The initial render model; defaults to <see cref="SummaryStatsRowModel.Pending"/>.</param>
    /// <param name="diagnostics">The PII-safe diagnostics sink for the <c>view.opened</c> event.</param>
    public SummaryStatsRow(
        ILocalizer localizer,
        SummaryStatsRowModel? model = null,
        SummaryStatsRowDiagnostics? diagnostics = null)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        _localizer = localizer;
        _model = model ?? SummaryStatsRowModel.Pending;
        _diagnostics = diagnostics ?? new SummaryStatsRowDiagnostics();

        IsTabStop = false;
        HorizontalContentAlignment = HorizontalAlignment.Stretch;
        VerticalContentAlignment = VerticalAlignment.Stretch;

        Loaded += OnLoaded;
        SizeChanged += OnSizeChanged;
        Render();
    }

    /// <summary>The canonical diagnostics slug this surface registers under (<c>SummaryStatsRow</c>).</summary>
    public static string Slug => SummaryStatsRowRegistration.Slug;

    /// <summary>The render model; reassigning re-projects and re-renders the surface.</summary>
    public SummaryStatsRowModel Model
    {
        get => _model;
        set
        {
            ArgumentNullException.ThrowIfNull(value);
            _model = value;
            Render();
        }
    }

    private void OnLoaded(object sender, RoutedEventArgs e)
    {
        if (_opened)
        {
            return;
        }

        _opened = true;
        _diagnostics.RecordViewOpened();
    }

    private void OnSizeChanged(object sender, SizeChangedEventArgs e)
    {
        int desired = ColumnsForWidth(e.NewSize.Width);
        if (desired != _columns)
        {
            _columns = desired;
            Render();
        }
    }

    private static int ColumnsForWidth(double width) => width switch
    {
        // web: grid-cols-1 sm:grid-cols-2 lg:grid-cols-4.
        <= 0 => SummaryStatsRowProjection.TileCount,
        < SmBreakpoint => 1,
        < LgBreakpoint => 2,
        _ => SummaryStatsRowProjection.TileCount,
    };

    private void Render()
    {
        var display = SummaryStatsRowProjection.Project(_model, _localizer, DateTimeOffset.Now);
        AutomationProperties.SetName(this, display.AutomationName);

        Content = display.State == SummaryStatsRowState.Loading
            ? BuildLoading(display)
            : BuildReady(display);
    }

    // ── Loading (parent still fetching the security history) ─────────────────────────────────────────────
    private Grid BuildLoading(SummaryStatsRowDisplay display)
    {
        var grid = BuildGrid();
        for (int i = 0; i < SummaryStatsRowProjection.TileCount; i++)
        {
            var tile = new TsSkeleton
            {
                BlockHeight = SkeletonTileHeight,
                ReduceMotion = MotionPreference.ReduceMotion,
                HorizontalAlignment = HorizontalAlignment.Stretch,
            };
            Place(grid, tile, i);
        }

        AutomationProperties.SetName(grid, display.LoadingLabel);
        LiveRegion.Configure(grid);
        LiveRegion.Announce(grid);
        return grid;
    }

    // ── Ready (the web render: FadeIn over the four metric tiles) ────────────────────────────────────────
    private TsFadeIn BuildReady(SummaryStatsRowDisplay display)
    {
        var grid = BuildGrid();
        for (int i = 0; i < display.Cards.Count; i++)
        {
            Place(grid, BuildTile(display.Cards[i]), i);
        }

        return new TsFadeIn
        {
            Content = grid,
            HorizontalAlignment = HorizontalAlignment.Stretch,
            VerticalAlignment = VerticalAlignment.Top,
        };
    }

    private static TsMetricCard BuildTile(SummaryStat card)
    {
        var tile = new TsMetricCard
        {
            Label = card.Label,
            Value = card.Value,
            AccentBrushKey = card.AccentBrushKey,
            HorizontalAlignment = HorizontalAlignment.Stretch,
        };
        AutomationProperties.SetName(tile, card.AutomationName);
        return tile;
    }

    private Grid BuildGrid()
    {
        var grid = new Grid { ColumnSpacing = TileGap, RowSpacing = TileGap };
        for (int c = 0; c < _columns; c++)
        {
            grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        }

        int rows = (int)Math.Ceiling(SummaryStatsRowProjection.TileCount / (double)_columns);
        for (int r = 0; r < rows; r++)
        {
            grid.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto });
        }

        return grid;
    }

    private void Place(Grid grid, FrameworkElement tile, int index)
    {
        Grid.SetColumn(tile, index % _columns);
        Grid.SetRow(tile, index / _columns);
        grid.Children.Add(tile);
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/SummaryStatsRow/SummaryStatsRow.kt
// The native Jetpack Compose + Material 3 SummaryStatsRow feature view — a parity port of
// web/src/features/admin/components/security-access/SummaryStatsRow.tsx. The web component renders four
// summary MetricCards (current secure/unsecure status, the relative age of the last lock change, the sentry
// uptime percentage, and the total security-event count) in a responsive 1 / 2 / 4-column grid, switching
// the whole grid to four skeleton tiles while the owning page's query loads. This port keeps that contract:
// the four cards always render (showing "—" and zeros, never a blank box, when a value is absent), the grid
// reflows at the web Tailwind `sm` (640dp) and `lg` (1024dp) breakpoints, and the resolved grid fades in
// exactly as the web `<FadeIn>` wrapper does.
//
// Every derivation flows through the pure [SummaryStatsRowProjection]; the composable is a thin render
// layer that resolves the i18n labels (P1/S10) and the design-token accents (P1/S9) and hands them to the
// shared MetricCard. The card labels, the Secure/Unsecure values, and the relative-age microcopy all
// resolve through the generated catalog (`admin.security.*` + `freshness.*` keys) — there is no English
// literal in this file. The one-shot `view.opened` diagnostic (P1/S11) is emitted on first composition.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/feature-views/SummaryStatsRow) cannot form a valid Kotlin package.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.summarystatsrow

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.graphics.vector.PathBuilder
import androidx.compose.ui.graphics.vector.path
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.teslasync.android.R
import io.teslasync.android.components.datadisplay.DataDisplayGlyphs
import io.teslasync.android.components.datadisplay.MetricCard
import io.teslasync.android.components.feedback.Skeleton
import io.teslasync.android.components.motion.FadeIn
import io.teslasync.android.data.LocalDataContainer
import io.teslasync.android.ui.theme.TeslaSyncTheme
import io.teslasync.android.ui.theme.TeslaTokens
import io.teslasync.android.ui.theme.generated.Spacing
import io.teslasync.shared.core.diagnostics.Logger
import java.util.Locale

/** Web Tailwind `lg` breakpoint (1024px): at or above this width the four cards lay out in a single row. */
private val GRID_LG_MIN_WIDTH: Dp = 1024.dp

/** Web Tailwind `sm` breakpoint (640px): at or above this width the cards lay out two-per-row. */
private val GRID_SM_MIN_WIDTH: Dp = 640.dp

private const val GRID_COLUMNS_LG: Int = 4
private const val GRID_COLUMNS_SM: Int = 2
private const val GRID_COLUMNS_BASE: Int = 1

/** The four summary tiles, matching the web component's fixed card set. */
private const val CARD_COUNT: Int = 4

/** Web `<Skeleton height={88} />` — each loading tile is 88dp tall. */
private val SKELETON_HEIGHT: Dp = 88.dp

/**
 * Stateful entry point — the faithful 1:1 port of the web `SummaryStatsRow({ isSecure, lastLockChange,
 * sentryUptime, totalEvents, isLoading })` props. Records the one-shot `view.opened` diagnostic on first
 * composition (P1/S11), projects the props onto a [SummaryStatsRowDisplay] via the pure
 * [SummaryStatsRowProjection], and renders.
 *
 * @param lastLockChange ISO-8601 timestamp of the most recent lock change, or `null` when unknown.
 * @param nowMillis wall clock the last-lock relative age is measured against; defaults to the system clock
 *   and is injectable so previews/tests stay deterministic.
 * @param logger the sanctioned redacting logger; defaults to the app's `LocalDataContainer`.
 */
@Composable
fun SummaryStatsRow(
    isSecure: Boolean,
    lastLockChange: String?,
    sentryUptime: Double,
    totalEvents: Int,
    isLoading: Boolean,
    modifier: Modifier = Modifier,
    nowMillis: Long = System.currentTimeMillis(),
    logger: Logger = LocalDataContainer.current.logger,
) {
    LaunchedEffect(Unit) { SummaryStatsRowDiagnostics.recordViewOpened(logger) }
    val display =
        remember(isSecure, lastLockChange, sentryUptime, totalEvents, isLoading, nowMillis) {
            SummaryStatsRowProjection.project(
                summary =
                    SecuritySummary(
                        isSecure = isSecure,
                        lastLockChange = lastLockChange,
                        sentryUptime = sentryUptime,
                        totalEvents = totalEvents,
                    ),
                loading = isLoading,
                nowMillis = nowMillis,
            )
        }
    SummaryStatsRowContent(display = display, modifier = modifier)
}

/**
 * Stateless renderer — the unit/UI-test and preview entry point. While [SummaryStatsRowDisplay.loading] is
 * true it shows the four-tile skeleton grid (web `isLoading` branch); otherwise it fades in the four
 * MetricCards. Every card is always present and always carries an accessible label + value (the last-lock
 * card resolves to "—" when the timestamp is absent), so no surface is ever hidden or blank.
 */
@Composable
fun SummaryStatsRowContent(
    display: SummaryStatsRowDisplay,
    modifier: Modifier = Modifier,
) {
    if (display.loading) {
        SummaryStatsLoading(modifier = modifier)
        return
    }

    val locale: Locale = LocalConfiguration.current.locales[0]
    val statusLabel = stringResource(R.string.translation_admin_security_stat_status)
    val secureValue = stringResource(R.string.translation_admin_security_secure)
    val unsecureValue = stringResource(R.string.translation_admin_security_unsecure)
    val lastLockLabel = stringResource(R.string.translation_admin_security_stat_lastLock)
    val sentryLabel = stringResource(R.string.translation_admin_security_stat_sentryUptime)
    val totalLabel = stringResource(R.string.translation_admin_security_stat_totalEvents)
    val lockLabels =
        LockChangeLabels(
            dash = EM_DASH,
            justNow = stringResource(R.string.translation_freshness_justNow),
            minutesAgo = stringResource(R.string.translation_freshness_minutes),
            hoursAgo = stringResource(R.string.translation_freshness_hours),
            daysAgo = stringResource(R.string.translation_freshness_days),
        )

    // Web `MetricCard` color → design-token accent: secure/unsecure are the success/danger semantics; the
    // web 'cyan' is the info token (both #00F0FF), and the web 'blue'/'purple' map to the brand chart
    // palette's speed/power hues (#3B82F6 / #A855F7, the latter an exact match for web's neon-purple).
    val statusValue = if (display.isSecure) secureValue else unsecureValue
    val statusAccent = if (display.isSecure) TeslaTokens.status.success else TeslaTokens.status.danger
    val lastLockValue = SummaryStatsRowProjection.formatLockChange(display.lastLock, lockLabels)
    val sentryValue = SummaryStatsRowProjection.formatUptimePercent(display.sentryUptime, locale)
    val totalValue = SummaryStatsRowProjection.formatEventCount(display.totalEvents)

    FadeIn(modifier = modifier) {
        SummaryStatsGrid(
            cards =
                listOf(
                    { cardModifier ->
                        MetricCard(
                            label = statusLabel,
                            value = statusValue,
                            icon = SummaryStatsRowGlyphs.ShieldCheck,
                            accent = statusAccent,
                            modifier = cardModifier,
                        )
                    },
                    { cardModifier ->
                        MetricCard(
                            label = lastLockLabel,
                            value = lastLockValue,
                            icon = DataDisplayGlyphs.Clock,
                            accent = TeslaTokens.status.info,
                            modifier = cardModifier,
                        )
                    },
                    { cardModifier ->
                        MetricCard(
                            label = sentryLabel,
                            value = sentryValue,
                            icon = SummaryStatsRowGlyphs.Activity,
                            accent = TeslaTokens.chart.speed,
                            modifier = cardModifier,
                        )
                    },
                    { cardModifier ->
                        MetricCard(
                            label = totalLabel,
                            value = totalValue,
                            icon = SummaryStatsRowGlyphs.BarChart3,
                            accent = TeslaTokens.chart.power,
                            modifier = cardModifier,
                        )
                    },
                ),
        )
    }
}

/**
 * The loading branch — four skeleton tiles in the same responsive grid as the resolved cards (web
 * `Array.from({ length: 4 }).map(... <Skeleton height={88} />)`). The grid carries a single TalkBack
 * "Loading" content description so the loading state is announced rather than read as four empty boxes.
 */
@Composable
private fun SummaryStatsLoading(modifier: Modifier = Modifier) {
    val loadingLabel = stringResource(R.string.translation_a11y_loading)
    val skeleton: @Composable (Modifier) -> Unit = { cardModifier ->
        Skeleton(modifier = cardModifier, height = SKELETON_HEIGHT)
    }
    SummaryStatsGrid(
        modifier = modifier.semantics { contentDescription = loadingLabel },
        cards = List(CARD_COUNT) { skeleton },
    )
}

/**
 * Lays out the [cards] as the web responsive grid: four-per-row at or above [GRID_LG_MIN_WIDTH] (`lg:4`),
 * two-per-row at or above [GRID_SM_MIN_WIDTH] (`sm:2`), and stacked below it (`default:1`). Each card fills
 * its column via [Modifier.weight]; a partial trailing row is padded with weighted spacers so the cards keep
 * a uniform width. Card cells are spaced by `Spacing.md`, the native expression of the web `gap-4`.
 */
@Composable
private fun SummaryStatsGrid(
    cards: List<@Composable (Modifier) -> Unit>,
    modifier: Modifier = Modifier,
) {
    BoxWithConstraints(modifier = modifier.fillMaxWidth()) {
        val columns =
            when {
                maxWidth >= GRID_LG_MIN_WIDTH -> GRID_COLUMNS_LG
                maxWidth >= GRID_SM_MIN_WIDTH -> GRID_COLUMNS_SM
                else -> GRID_COLUMNS_BASE
            }
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(Spacing.md),
        ) {
            cards.chunked(columns).forEach { rowCards ->
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(Spacing.md),
                ) {
                    rowCards.forEach { card -> card(Modifier.weight(1f)) }
                    repeat(columns - rowCards.size) { Spacer(modifier = Modifier.weight(1f)) }
                }
            }
        }
    }
}

/**
 * The three glyphs this surface needs that the shared [DataDisplayGlyphs] set does not carry. The web uses
 * lucide `ShieldCheck`, `Activity`, and `BarChart3`; Android ships no equivalents without the frozen
 * `material-icons-extended` artifact, so — exactly as the shared sets do for their lucide ports — they are
 * authored here as 24×24 stroked vectors faithful to the lucide paths. The `Clock` glyph (last-lock card)
 * is reused from `DataDisplayGlyphs`.
 */
private object SummaryStatsRowGlyphs {
    val ShieldCheck: ImageVector =
        stroked("ShieldCheck") {
            moveTo(12f, 3f)
            lineTo(19f, 6f)
            lineTo(19f, 12f)
            curveTo(19f, 16.5f, 16f, 19.5f, 12f, 21f)
            curveTo(8f, 19.5f, 5f, 16.5f, 5f, 12f)
            lineTo(5f, 6f)
            close()
            moveTo(9f, 12.5f)
            lineTo(11f, 14.5f)
            lineTo(15.5f, 9.5f)
        }

    val Activity: ImageVector =
        stroked("Activity") {
            moveTo(22f, 12f)
            lineTo(18f, 12f)
            lineTo(15f, 21f)
            lineTo(9f, 3f)
            lineTo(6f, 12f)
            lineTo(2f, 12f)
        }

    val BarChart3: ImageVector =
        stroked("BarChart3") {
            moveTo(3f, 3f)
            lineTo(3f, 21f)
            lineTo(21f, 21f)
            moveTo(8f, 17f)
            lineTo(8f, 14f)
            moveTo(13f, 17f)
            lineTo(13f, 5f)
            moveTo(18f, 17f)
            lineTo(18f, 9f)
        }

    private fun stroked(
        name: String,
        build: PathBuilder.() -> Unit,
    ): ImageVector =
        ImageVector
            .Builder(
                name = name,
                defaultWidth = 24.dp,
                defaultHeight = 24.dp,
                viewportWidth = 24f,
                viewportHeight = 24f,
            ).apply {
                path(
                    stroke = SolidColor(Color.Black),
                    strokeLineWidth = 2f,
                    strokeLineCap = StrokeCap.Round,
                    strokeLineJoin = StrokeJoin.Round,
                    pathBuilder = build,
                )
            }.build()
}

// ── Previews (tooling-only; @Preview entry points exercise each render branch) ──────────────────────

@Preview(name = "Loading", showBackground = true)
@Composable
private fun SummaryStatsRowLoadingPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        SummaryStatsRowContent(
            SummaryStatsRowDisplay(
                loading = true,
                isSecure = true,
                lastLock = LockChangeAge.Unknown,
                sentryUptime = 0.0,
                totalEvents = 0,
            ),
        )
    }
}

@Preview(name = "Resolved — secure", showBackground = true)
@Composable
private fun SummaryStatsRowSecurePreview() {
    TeslaSyncTheme(dynamicColor = false) {
        SummaryStatsRowContent(
            SummaryStatsRowDisplay(
                loading = false,
                isSecure = true,
                lastLock = LockChangeAge.Minutes(5),
                sentryUptime = 98.0,
                totalEvents = 1234,
            ),
        )
    }
}

@Preview(name = "Resolved — unsecure", showBackground = true)
@Composable
private fun SummaryStatsRowUnsecurePreview() {
    TeslaSyncTheme(dynamicColor = false) {
        SummaryStatsRowContent(
            SummaryStatsRowDisplay(
                loading = false,
                isSecure = false,
                lastLock = LockChangeAge.Hours(3),
                sentryUptime = 42.0,
                totalEvents = 87,
            ),
        )
    }
}

@Preview(name = "Resolved — empty (no lock time / zeros)", showBackground = true)
@Composable
private fun SummaryStatsRowEmptyPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        SummaryStatsRowContent(
            SummaryStatsRowDisplay(
                loading = false,
                isSecure = true,
                lastLock = LockChangeAge.Unknown,
                sentryUptime = 0.0,
                totalEvents = 0,
            ),
        )
    }
}
